pkg/containerruntime/containerd.go

```diff
@@ -172,3 +170,132 @@ systemctl enable --now containerd
+	SandboxImage string                 `toml:"sandbox_image,omitempty"`
+}
+
+type containerdCRISettings struct {
+	Runtimes map[string]containerdCRIRuntime `toml:"runtimes"`
+}
+
+type containerdCRIRuntime struct {
+	RuntimeType string      `toml:"runtime_type"`
+	Options     interface{} `toml:"options"`
+}
+
+type containerdCRIRuncOptions struct {
+	SystemdCgroup bool
+}
+
+type containerdCRIRegistry struct {
+	Mirrors map[string]containerdRegistryMirror `toml:"mirrors"`
+	Configs map[string]containerdRegistryConfig `toml:"configs"`
+}
+
+type containerdRegistryMirror struct {
+	Endpoint []string `toml:"endpoint"`
+}
+
+type containerdRegistryConfig struct {
+	TLS  *containerdRegistryTLSConfig `toml:"tls"`
+	Auth *AuthConfig                  `toml:"auth"`
+}
+
+type containerdRegistryTLSConfig struct {
+	InsecureSkipVerify bool `toml:"insecure_skip_verify"`
+}
+
+// AuthConfig is a COPY of github.com/containerd/containerd/pkg/cri/config.AuthConfig.
+// AuthConfig contains the config related to authentication to a specific registry
+type AuthConfig struct {
+	// Username is the username to login the registry.
+	Username string `toml:"username,omitempty" json:"username,omitempty"`
+	// Password is the password to login the registry.
+	Password string `toml:"password,omitempty" json:"password,omitempty"`
+	// Auth is a base64 encoded string from the concatenation of the username,
+	// a colon, and the password.
+	Auth string `toml:"auth,omitempty" json:"auth,omitempty"`
+	// IdentityToken is used to authenticate the user and get
+	// an access token for the registry.
+	IdentityToken string `toml:"identitytoken,omitempty" json:"identitytoken,omitempty"`
+}
+
+func (eng *Containerd) Config() (string, error) {
+	criPlugin := containerdCRIPlugin{
+		SandboxImage: eng.sandboxImage,
+		Containerd: &containerdCRISettings{
+			Runtimes: map[string]containerdCRIRuntime{
+				"runc": {
+					RuntimeType: "io.containerd.runc.v2",
+					Options: containerdCRIRuncOptions{
+						SystemdCgroup: true,
+					},
+				},
+			},
+		},
+		Registry: &containerdCRIRegistry{
+			Mirrors: map[string]containerdRegistryMirror{
+				"docker.io": {
+					Endpoint: []string{"https://registry-1.docker.io"},
+				},
+			},
+		},
+	}
+
+	for registryName := range eng.registryMirrors {
+		registry := criPlugin.Registry.Mirrors[registryName]
+		registry.Endpoint = eng.registryMirrors[registryName]
+		criPlugin.Registry.Mirrors[registryName] = registry
+	}
+
+	if len(eng.insecureRegistries) != 0 || len(eng.registryCredentials) != 0 {
+		criPlugin.Registry.Configs = map[string]containerdRegistryConfig{}
+	}
+
+	for _, registry := range eng.insecureRegistries {
+		criPlugin.Registry.Configs[registry] = containerdRegistryConfig{
+			TLS: &containerdRegistryTLSConfig{
+				InsecureSkipVerify: true,
+			},
+		}
+	}
+
+	for registry, auth := range eng.registryCredentials {
+		regConfig := criPlugin.Registry.Configs[registry]
+		auth := auth
```

Comment:
What’s the point of this line? To avoid lint failures?

Reply:
to make a copy, because original auth is an iteration variable, and it's being reused every iteration. We need a copy because we take pointer of this var inside the loop.